File: ognibuild/dist.py
#!/usr/bin/python3
# Copyright (C) 2020 Jelmer Vernooij <jelmer@jelmer.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

__all__ = [
    "UnidentifiedError",
    "DetailedFailure",
    "run_dist",
    "create_dist_schroot",
    "create_dist",
    "dist",
]

import errno
from functools import partial
import logging
import os
import sys
from typing import Optional, List

from breezy.tree import Tree
from breezy.workingtree import WorkingTree

from buildlog_consultant.common import (
    NoSpaceOnDevice,
)

from debian.deb822 import Deb822


from . import DetailedFailure, UnidentifiedError
from .dist_catcher import DistNoTarball
from .fix_build import iterate_with_build_fixers
from .logs import LogManager, NoLogManager
from .buildsystem import NoBuildToolsFound
from .resolver import auto_resolver
from .session import Session
from .session.schroot import SchrootSession


DIST_LOG_FILENAME = 'dist.log'


def run_dist(session, buildsystems, resolver, fixers, target_directory,
             quiet=False, log_manager=None):
    # Some things want to write to the user's home directory,
    # e.g. pip caches in ~/.cache
    session.create_home()

    logging.info('Using dependency resolver: %s', resolver)

    if log_manager is None:
        log_manager = NoLogManager()

    for buildsystem in buildsystems:
        return iterate_with_build_fixers(fixers, log_manager.wrap(
            partial(buildsystem.dist, session, resolver, target_directory,
                    quiet=quiet)))

    raise NoBuildToolsFound()


def dist(session, export_directory, reldir, target_dir, log_manager, *,
         version: Optional[str] = None, quiet=False):
    from .fix_build import BuildFixer
    from .buildsystem import detect_buildsystems
    from .buildlog import InstallFixer
    from .fixers import (
        GitIdentityFixer,
        MissingGoSumEntryFixer,
        SecretGpgKeyFixer,
        UnexpandedAutoconfMacroFixer,
        GnulibDirectoryFixer,
    )

    if version:
        # TODO(jelmer): Shouldn't include backend-specific code here
        os.environ['SETUPTOOLS_SCM_PRETEND_VERSION'] = version

    # TODO(jelmer): use scan_buildsystems to also look in subdirectories
    buildsystems = list(detect_buildsystems(export_directory))
    resolver = auto_resolver(session)
    fixers: List[BuildFixer] = [
        UnexpandedAutoconfMacroFixer(session, resolver),
        GnulibDirectoryFixer(session),
        MissingGoSumEntryFixer(session),
        InstallFixer(resolver)]

    if session.is_temporary:
        # Only muck about with temporary sessions
        fixers.extend([
            GitIdentityFixer(session),
            SecretGpgKeyFixer(session),
        ])

    session.chdir(reldir)

    # Some things want to write to the user's home directory,
    # e.g. pip caches in ~/.cache
    session.create_home()

    logging.info('Using dependency resolver: %s', resolver)

    for buildsystem in buildsystems:
        return iterate_with_build_fixers(fixers, log_manager.wrap(
            partial(
                buildsystem.dist, session, resolver, target_dir,
                quiet=quiet)))

    raise NoBuildToolsFound()


# This is the function used by debianize()
def create_dist(
    session: Session,
    tree: Tree,
    target_dir: str,
    include_controldir: bool = True,
    subdir: Optional[str] = None,
    log_manager: Optional[LogManager] = None,
    version: Optional[str] = None,
) -> Optional[str]:
    """Create a dist tarball for a tree.

    Args:
      session: session to run it
      tree: Tree object to work in
      target_dir: Directory to write tarball into
      include_controldir: Whether to include the version control directory
      subdir: subdirectory in the tree to operate in
    """
    if subdir is None:
        subdir = "package"
    try:
        export_directory, reldir = session.setup_from_vcs(
            tree, include_controldir=include_controldir, subdir=subdir
        )
    except OSError as e:
        if e.errno == errno.ENOSPC:
            raise DetailedFailure(1, ["mkdtemp"], NoSpaceOnDevice()) from e
        raise

    if log_manager is None:
        log_manager = NoLogManager()

    return dist(session, export_directory, reldir, target_dir,
                log_manager=log_manager, version=version)


def create_dist_schroot(
    tree: Tree,
    target_dir: str,
    chroot: str,
    packaging_tree: Optional[Tree] = None,
    packaging_subpath: Optional[str] = None,
    include_controldir: bool = True,
    subdir: Optional[str] = None,
    log_manager: Optional[LogManager] = None,
) -> Optional[str]:
    """Create a dist tarball for a tree.

    Args:
      session: session to run it
      tree: Tree object to work in
      target_dir: Directory to write tarball into
      include_controldir: Whether to include the version control directory
      subdir: subdirectory in the tree to operate in
    """
    with SchrootSession(chroot) as session:
        if packaging_tree is not None:
            from .debian import satisfy_build_deps

            satisfy_build_deps(session, packaging_tree, packaging_subpath)
        return create_dist(
                session, tree, target_dir,
                include_controldir=include_controldir, subdir=subdir,
                log_manager=log_manager)


def main(argv=None):
    import argparse
    import breezy.bzr  # noqa: F401
    import breezy.git  # noqa: F401
    from breezy.export import export

    parser = argparse.ArgumentParser(argv)
    parser.add_argument(
        "--chroot",
        default="unstable-amd64-sbuild",
        type=str,
        help="Name of chroot to use",
    )
    parser.add_argument(
        "directory",
        default=".",
        type=str,
        nargs="?",
        help="Directory with upstream source.",
    )
    parser.add_argument(
        "--packaging-directory", type=str, help="Path to packaging directory."
    )
    parser.add_argument(
        "--target-directory", type=str, default="..", help="Target directory"
    )
    parser.add_argument("--verbose", action="store_true", help="Be verbose")
    parser.add_argument("--mode", choices=["auto", "vcs", "buildsystem"],
                        type=str,
                        help="Mechanism to use to create buildsystem")
    parser.add_argument(
        "--include-controldir", action="store_true",
        help="Clone rather than export."
    )

    args = parser.parse_args()

    if args.verbose:
        logging.basicConfig(level=logging.DEBUG, format="%(message)s")
    else:
        logging.basicConfig(level=logging.INFO, format="%(message)s")

    tree = WorkingTree.open(args.directory)

    packaging_tree: Optional[WorkingTree]
    subdir: Optional[str]

    if args.packaging_directory:
        packaging_tree = WorkingTree.open(args.packaging_directory)
        with packaging_tree.lock_read():  # type: ignore
            source = Deb822(
                packaging_tree.get_file("debian/control"))  # type: ignore
        package = source["Source"]
        subdir = package
    else:
        packaging_tree = None
        subdir = None

    if args.mode == 'vcs':
        export(tree, "dist.tar.gz", "tgz", None)
    elif args.mode in ('auto', 'buildsystem'):
        try:
            ret = create_dist_schroot(
                tree,
                subdir=subdir,
                target_dir=os.path.abspath(args.target_directory),
                packaging_tree=packaging_tree,
                chroot=args.chroot,
                include_controldir=args.include_controldir,
            )
        except NoBuildToolsFound:
            if args.mode == 'buildsystem':
                logging.fatal('No build tools found, unable to create tarball')
                return 1
            logging.info(
                "No build tools found, falling back to simple export.")
            export(tree, "dist.tar.gz", "tgz", None)
        except NotImplementedError:
            if args.mode == 'buildsystem':
                logging.fatal('Unable to ask buildsystem for tarball')
                return 1
            logging.info(
                "Build system does not support dist tarball creation, "
                "falling back to simple export."
            )
            export(tree, "dist.tar.gz", "tgz", None)
        except UnidentifiedError as e:
            logging.fatal("Unidentified error: %r", e.lines)
            return 1
        except DetailedFailure as e:
            logging.fatal("Identified error during dist creation: %s", e.error)
            return 1
        except DistNoTarball:
            logging.fatal("dist operation did not create a tarball")
            return 1
        else:
            logging.info("Created %s", ret)
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: src/dist.rs
use crate::buildsystem::{detect_buildsystems, Error};
use crate::fix_build::{iterate_with_build_fixers, BuildFixer, InterimError};
use crate::fixers::*;
use crate::installer::{
    auto_installation_scope, auto_installer, Error as InstallerError, InstallationScope,
};
use crate::logs::{wrap, LogManager};
use crate::session::Session;
use std::ffi::OsString;
use std::path::Path;

/// Create a distribution package using the detected build system.
///
/// # Arguments
/// * `session` - The session to run commands in
/// * `export_directory` - Directory to search for build systems
/// * `reldir` - Relative directory to change to before building
/// * `target_dir` - Directory to write distribution package to
/// * `log_manager` - Log manager for capturing output
/// * `version` - Optional version to use for the package
/// * `quiet` - Whether to suppress output
///
/// # Returns
/// The filename of the created distribution package
pub fn dist(
    session: &mut dyn Session,
    export_directory: &Path,
    reldir: &Path,
    target_dir: &Path,
    log_manager: &mut dyn LogManager,
    version: Option<&str>,
    quiet: bool,
) -> Result<OsString, Error> {
    session.chdir(reldir)?;

    if let Some(version) = version {
        // TODO(jelmer): Shouldn't include backend-specific code here
        std::env::set_var("SETUPTOOLS_SCM_PRETEND_VERSION", version);
    }

    // TODO(jelmer): use scan_buildsystems to also look in subdirectories
    let buildsystems = detect_buildsystems(export_directory);
    let scope = auto_installation_scope(session);
    let installer = auto_installer(session, scope, None);
    let mut fixers: Vec<Box<dyn BuildFixer<InstallerError>>> = vec![
        Box::new(UnexpandedAutoconfMacroFixer::new(
            session,
            installer.as_ref(),
        )),
        Box::new(GnulibDirectoryFixer::new(session)),
        Box::new(MinimumAutoconfFixer::new(session)),
        Box::new(MissingGoSumEntryFixer::new(session)),
        Box::new(InstallFixer::new(
            installer.as_ref(),
            InstallationScope::User,
        )),
    ];

    if session.is_temporary() {
        // Only muck about with temporary sessions
        fixers.extend([
            Box::new(GitIdentityFixer::new(session)) as Box<dyn BuildFixer<InstallerError>>,
            Box::new(SecretGpgKeyFixer::new(session)) as Box<dyn BuildFixer<InstallerError>>,
        ]);
    }

    // Some things want to write to the user's home directory, e.g. pip caches in ~/.cache
    session.create_home()?;

    for buildsystem in buildsystems {
        return Ok(iterate_with_build_fixers(
            fixers
                .iter()
                .map(|x| x.as_ref())
                .collect::<Vec<_>>()
                .as_slice(),
            || -> Result<_, InterimError<Error>> {
                Ok(wrap(log_manager, || -> Result<_, Error> {
                    buildsystem.dist(session, installer.as_ref(), target_dir, quiet)
                })?)
            },
            None,
        )?);
    }

    Err(Error::NoBuildSystemDetected)
}

#[cfg(feature = "breezy")]
// This is the function used by debianize()
/// Create a dist tarball for a tree.
///
/// # Arguments
/// * `session` - session to run it
/// * `tree` - Tree object to work in
/// * `target_dir` - Directory to write tarball into
/// * `include_controldir` - Whether to include the version control directory
/// * `temp_subdir` - name of subdirectory in which to check out the source code;
///            defaults to "package"
pub fn create_dist<T: crate::vcs::DupableTree>(
    session: &mut dyn Session,
    tree: &T,
    target_dir: &Path,
    include_controldir: Option<bool>,
    log_manager: &mut dyn LogManager,
    version: Option<&str>,
    _subpath: &Path,
    temp_subdir: Option<&str>,
) -> Result<OsString, Error> {
    let temp_subdir = temp_subdir.unwrap_or("package");

    let project = session.project_from_vcs(tree, include_controldir, Some(temp_subdir))?;

    dist(
        session,
        project.external_path(),
        project.internal_path(),
        target_dir,
        log_manager,
        version,
        false,
    )
}

#[cfg(feature = "breezy")]
#[cfg(target_os = "linux")]
/// Create a dist tarball for a tree.
///
/// # Arguments
/// * `session` - session to run it
/// * `tree` - Tree object to work in
/// * `target_dir` - Directory to write tarball into
/// * `include_controldir` - Whether to include the version control directory
/// * `temp_subdir` - name of subdirectory in which to check out the source code;
///             defaults to "package"
pub fn create_dist_schroot<T: crate::vcs::DupableTree>(
    tree: &T,
    target_dir: &Path,
    chroot: &str,
    packaging_tree: Option<&dyn breezyshim::tree::Tree>,
    packaging_subpath: Option<&Path>,
    include_controldir: Option<bool>,
    subpath: &Path,
    log_manager: &mut dyn LogManager,
    version: Option<&str>,
    temp_subdir: Option<&str>,
) -> Result<OsString, Error> {
    // TODO(jelmer): pass in package name as part of session prefix
    let mut session = crate::session::schroot::SchrootSession::new(chroot, Some("ognibuild-dist"))?;
    #[cfg(feature = "debian")]
    if let (Some(packaging_tree), Some(packaging_subpath)) = (packaging_tree, packaging_subpath) {
        crate::debian::satisfy_build_deps(&session, packaging_tree, packaging_subpath)
            .map_err(|e| Error::Other(format!("Failed to satisfy build dependencies: {:?}", e)))?;
    }
    #[cfg(not(feature = "debian"))]
    if packaging_tree.is_some() || packaging_subpath.is_some() {
        log::warn!("Ignoring packaging tree and subpath as debian feature is not enabled");
    }
    create_dist(
        &mut session,
        tree,
        target_dir,
        include_controldir,
        log_manager,
        version,
        subpath,
        temp_subdir,
    )
}
